<h1 style="text-align: center;">Desafío - Funciones y probabilidades</h1>

- Para realizar este desafío debes haber estudiado previamente todo el material disponible correspondiente a la unidad.
- Una vez terminado el desafío, comprime la carpeta que contiene el desarrollo de los requerimientos solicitados y sube el .zip en el LMS.
- Para la realización del desafío necesitarás apoyarte del archivo `Apoyo Desafío - Calculando probabilidades.zip`, disponible en la <a href="https://empieza.desafiolatam.com/cohorts/2091/sections/78268/activities/119096/items/610788">plataforma</a>.
- Desarrollo desafío:
    - El desafío se debe desarrollar de manera Individual.
    - Puntaje total: 10 puntos.

## Desafío 1: Generación de funciones (1 punto)
- Genere funciones para calcular la media y varianza de un vector. Debe cumplir con los siguientes requisitos:
    - Ambas funciones deben ingresar un argumento x correspondiente al vector.
    - Las funciones deben contener docstrings con la documentación asociada a la variable y retornar un valor utilizando `return`.
- La función de la varianza se debe llamar a la función de la media.
- Utilice las funciones para reportar la información sobre `goles_favor`, `goles_contra` y `puntos`.

In [5]:
# Utiliza este bloque para definir las funciones solicitadas

def media(data_vector):
    ''' 
        Esta función retorna la media artimética del vector de datos ingresado como parámetro

        Parametros:
            a (list): una lista que contiene números a los cuales calcularles su media
        Retorno:
            media (float): El valor de la media calculada entre los números del vector ingresado
    '''
    acumulador = 0
    for i in data_vector:
        acumulador += i
    media = acumulador/len(data_vector)
    return media

def varianza(data_vector):
    ''' 
        Esta función retorna la varianza del vector de datos ingresado como parámetro

        Parametros:
            a (list): una lista que contiene números a los cuales calcularles su varianza
        Retorno:
            varianza (float): El valor de la varainza calculada entre los números del vector ingresado
    '''
    acumulador = 0
    valor_media = media(data_vector)
    for i in data_vector:
        acumulador += (i - valor_media)**2
    varianza = acumulador/(len(data_vector)-1)
    return varianza


7.5

In [158]:
# Utiliza este bloque para reportar la información sobre goles_favor, goles_contra y puntos
import pandas as pd

df = pd.read_csv('worldcup2014.csv')
media_goles_favor = media(df['goles_favor'])
media_goles_contra = media(df['goles_contra'])
varianza_goles_favor = round(varianza(df['goles_favor']),2)
varianza_goles_contra = round(varianza(df['goles_contra']),2)
media_puntos = media(df['puntos'])
varianza_puntos = round(varianza(df['puntos']),2)
print(f'El promedio de goles a favor de todos los equipos del mundial de fútbol es : {media_goles_favor} y su varianza es {varianza_goles_favor}')
print(f'El promedio de goles en contra de todos los equipos del mundial de fútbol es : {media_goles_contra} y su varianza es {varianza_goles_contra}')
print(f'El promedio de puntos de todos los equipos del mundial de fútbol es : {media_puntos} y su varianza es {varianza_puntos}')


El promedio de goles a favor de todos los equipos del mundial de fútbol es : 4.25 y su varianza es 5.35
El promedio de goles en contra de todos los equipos del mundial de fútbol es : 4.25 y su varianza es 4.97
El promedio de puntos de todos los equipos del mundial de fútbol es : 4.21875 y su varianza es 8.31


## Desafío 2: Uso de groupby y agg (2 puntos)
Utilizando el método `groupby` de la clase `DataFrame`, en conjunto con la función `.agg` de pandas, calcule la media, la varianza y desviación estándar de la cantidad de goles a favor, en contra y de la cantidad de puntos por continente.

In [159]:
# Utiliza esta celda para aplicar groupby y obtener las métricas solicitadas con agg
goles_favor = df.groupby(df['continent'])['goles_favor'].agg(['sum'])
goles_contra = df.groupby(df['continent'])['goles_contra'].agg(['sum'])
puntos = df.groupby(df['continent'])['puntos'].agg(['sum'])

media_gf = df.groupby(df['continent'])['goles_favor'].agg(['mean'])
std_gf = df.groupby(df['continent'])['goles_favor'].agg(['std'])
var_gf = df.groupby(df['continent'])['goles_favor'].agg(['var'])

media_gc = df.groupby(df['continent'])['goles_contra'].agg(['mean'])
std_gc =  df.groupby(df['continent'])['goles_contra'].agg(['std'])
var_gc = df.groupby(df['continent'])['goles_contra'].agg(['var'])

media_pt = df.groupby(df['continent'])['puntos'].agg(['mean'])
std_pt = df.groupby(df['continent'])['puntos'].agg(['std'])
var_pt = df.groupby(df['continent'])['puntos'].agg(['var'])

#df.groupby('continent')[['goles_favor', 'goles_contra','puntos']].agg(['sum','mean','var','std'])


In [156]:
# A continuación se muestran las cantidades asociadas a cada dato
print(f'Goles a favor: {goles_favor} \n')
print(f'Goles en contra: {goles_contra} \n')
print(f'Puntos por continente: {goles_contra} \n')


Goles a favor:               sum
continent        
africa         18
asia            9
europe         62
northamerica   13
southamerica   34 

Goles en contra:               sum
continent        
africa         28
asia           25
europe         52
northamerica   14
southamerica   17 

Puntos por continente:               sum
continent        
africa         28
asia           25
europe         52
northamerica   14
southamerica   17 



In [160]:
print(f'La media de los goles a favor por continente es: {media_gf} \n')
print(f'La media de los goles en contra por continente es: {media_gc} \n')
print(f'La media de los puntos por continente es: {media_pt} \n')


La media de los goles a favor por continente es:                   mean
continent             
africa        3.600000
asia          2.250000
europe        4.769231
northamerica  3.250000
southamerica  5.666667 

La media de los goles en contra por continente es:                   mean
continent             
africa        5.600000
asia          6.250000
europe        4.000000
northamerica  3.500000
southamerica  2.833333 

La media de los puntos por continente es:                   mean
continent             
africa        2.400000
asia          0.750000
europe        4.692308
northamerica  4.500000
southamerica  6.833333 



In [104]:
print(f'La varianza de los goles a favor por continente es: {var_gf} \n ')
print(f'La varianza de los goles en contra por continente es: {var_gc} \n ')
print(f'La varianza de los puntos por continente es: {var_pt} \n ')

La varianza de los goles a favor por continente es:                    std
continent             
africa        3.300000
asia          0.916667
europe        6.858974
northamerica  2.250000
southamerica  4.666667 
 
La varianza de los goles en contra por continente es:                     std
continent              
africa         4.800000
asia           4.250000
europe         3.833333
northamerica  11.000000
southamerica   0.566667 
 
La varianza de los puntos por continente es:                     std
continent              
africa         3.300000
asia           0.250000
europe         6.897436
northamerica  11.000000
southamerica   3.766667 
 


In [161]:
print(f'La desviación estándar de los goles a favor por continente es: {std_gf} \n')
print(f'La desviación estándar de los goles en contra por continente es: {std_gc} \n')
print(f'La desviación estándar de los puntos por continente es: {std_pt}')


La desviación estándar de los goles a favor por continente es:                    std
continent             
africa        1.816590
asia          0.957427
europe        2.618964
northamerica  1.500000
southamerica  2.160247 

La desviación estándar de los goles en contra por continente es:                    std
continent             
africa        2.190890
asia          2.061553
europe        1.957890
northamerica  3.316625
southamerica  0.752773 

La desviación estándar de los puntos por continente es:                    std
continent             
africa        1.816590
asia          0.500000
europe        2.626297
northamerica  3.316625
southamerica  1.940790


**¿En qué continente se observa una mayor cantidad de goles a favor?**

**Respuesta**: Europa

**¿En qué continente se observa una mayor cantidad de goles en contra?**

**Respuesta:** Europa

**¿En qué continente se observa una mayor cantidad de puntos en promedio?**

**Respuesta**: Suramérica

## Desafío 3: Simulaciones (3 puntos)

- Genere una función `generate_pet` que devuelva de forma aleatoria un string `'perro'` o `'gato'` un número `n` de veces. Ejecútela un par de veces.
    - Tip: Puede utilizar la función `np.random.choice` para retornar elementos al azar.
- Aplique la función generate_pet para generar 20 muestras.

In [137]:
# Use este bloque para definir la función generate_pet
import numpy as np
np.random.seed(2)

def generate_pet(n):
    '''
        Función que entrega una cantidad n de elementos en una lista de los cuales pueden tener el valor 'perro' o 'gato

        Parámetros:
            n (int): Cantidad de valores en la lista
        Retorno:
            Una lista con los valores 'perro' o 'gato'
    '''
    pets = ['perro', 'gato']
    return np.random.choice(pets,n)

In [138]:
# Use este bloque para generar las 20 muestras usando la función
df = pd.DataFrame({'pet':generate_pet(20)})
df['pet'].value_counts('%')

gato     0.6
perro    0.4
Name: pet, dtype: float64

**¿Cuál es la probabilidad de elegir un perro al azar? ¿Y un gato?**

**Respuesta:** No existe una probabilidad fija ya que depende de como se genere el resultado de la función generate_pet

**Agregue `np.random.seed(2)` al inicio del chunk. ¿Qué diferencia hay cuando se ejecuta la función varias veces luego de fijar la semilla?**

**Respuesta**: El generar una semilla previa, el resultado se hace predecible por lo que la probabilidad de elegir un perro es de : 0.4 y un fato de 0.6. El código es replicable

## Desafío 4: Función simuladora (4 puntos)
- Genere una función llamada `simulate_pets_prob` que tome como argumento un número finito de simulaciones a generar.
- La función debe simular dos situaciones `'young_pet'` y `'old_pet'`, y contar la ocurrencia de los siguientes casos:
    - De las dos mascotas simuladas (`'young_pet'` y `'old_pet'`), contar las ocasiones donde _por lo menos una_ de las mascotas sea un perro.
    - De las dos mascotas simuladas, contar las ocasiones donde `'old_pet'` sea un perro.
    - De las dos mascotas simuladas, contar las ocasiones donde los dos sean perros.
- El método debe tener una semilla pseudoaleatoria de 1.
- El output de la función debe entregarse en términos de Probabilidad.

In [183]:
# Use este bloque para definir la función
np.random.seed(1)
def simulate_pets_prob(n):
    '''
        Función que simula dos mascotas: old u young

        Parámetros:
            n (int): Cantidad de experimentos
        Return:
            La probabilidad de las ocurrencias de mascotas
    '''
    pets_age = pd.DataFrame([[generate_pet(1), generate_pet(1)] for i in range(n)], columns=['old_pet', 'young_pet']).applymap(lambda x: x[0])
    pets_age['is_dog'] = np.where((pets_age['young_pet'] == 'perro') | (pets_age['old_pet'] == 'perro'), 1, 0)
    pets_age['old_dog'] = np.where(pets_age['old_pet'] == 'perro', 1, 0)
    pets_age['both_dogs'] = np.where((pets_age['young_pet'] == 'perro') & (pets_age['old_pet'] == 'perro'), 1, 0)

    return pets_age.value_counts(normalize=True)


In [185]:
# Utilice este bloque para realizar el llamado
simulate_pets_prob().to_frame()


0
old_pet young_pet is_dog old_dog both_dogs       
perro   perro     1      1       1          0.265
gato    gato      0      0       0          0.260
        perro     1      0       0          0.240
perro   gato      1      1       0          0.235

**De los tres escenarios, ¿Cuál es el menos probable? ¿Cuál es el más probable? ¿Por qué?**

**Respuesta:**
El escenario menos probable es que un perro sea anciano y un gato jóven. Con una probabilidad de 0.235
El escenario más probable es que haya un par de perros jóven y anciano. Con una probabilidad de 0.265